core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -2448,29 +2448,28 @@ object Stream {
     private def zipWith_[O2, O3](that: Stream[F, O2])(
         k1: ZipWithCont[F, O, O3, Nothing],
         k2: ZipWithCont[F, O2, O3, Nothing])(f: (O, O2) => O3): Stream[F, O3] = {
-      def go(t1: (Segment[O, Unit], Stream[F, O]),
-             t2: (Segment[O2, Unit], Stream[F, O2])): Pull[F, O3, Option[Nothing]] =
-        (t1, t2) match {
-          case ((hd1, tl1), (hd2, tl2)) =>
-            Pull.segment(hd1.zipWith(hd2)(f)).flatMap {
-              case Left(((), extra2)) =>
-                tl1.pull.uncons.flatMap {
-                  case None      => k2(Left((extra2, tl2)))
-                  case Some(tl1) => go(tl1, (extra2, tl2))
-                }
-              case Right(((), extra1)) =>
-                tl2.pull.uncons.flatMap {
-                  case None      => k1(Left((extra1, tl1)))
-                  case Some(tl2) => go((extra1, tl1), tl2)
-                }
+      def go(leg1: StreamLeg[F, O], leg2: StreamLeg[F, O2]): Pull[F, O3, Option[Nothing]] =
+        Pull.segment(leg1.head.zipWith(leg2.head)(f)).flatMap {
+          case Left(((), extra2)) =>
+            leg1.uncons.flatMap {
+              case None      => k2(Left((extra2, leg2.stream)))
+              case Some(tl1) => go(tl1, leg2.setHead(extra2))
+            }
+          case Right(((), extra1)) =>
+            leg2.uncons.flatMap {
+              case None      => k1(Left((extra1, leg1.stream)))
+              case Some(tl2) => go(leg1.setHead(extra1), leg2)
```

Comment:
The issues with zips you are seeing may be because you accidentally used `leg2` here instead of `tl2`

Reply:
ok nice catch